atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,325 @@
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            awaitOrFail(manuallyFinishRunWith);
+            verifyNoInteractions(closeable);
+        }));
+
+        Future<?> refreshableFuture = executor.submit(() -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            settableRefreshable.update(UPDATED_VALUE);
+        });
+
+        Awaitility.await().during(Duration.ofSeconds(1)).until(() -> !refreshableFuture.isDone());
+        verifyNoInteractions(mockCloseable);
+
+        manuallyFinishRunWith.await();
+
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(refreshableFuture::isDone);
+        verify(mockCloseable).close();
+    }
+
+    @Test
+    public void refreshUpdatesResource() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseable));
+
+        settableRefreshable.update(UPDATED_VALUE);
+        container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseableExtra));
+    }
+
+    @Test
+    public void previousResourceClosedAfterRefresh() throws IOException {
+        // Creating a container has the side effect of managing all resources created within it.
+        getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        settableRefreshable.update(UPDATED_VALUE);
+        verify(mockCloseable).close();
+    }
+
+    @Test
+    public void latestResourceClosedAfterClose() throws IOException {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        settableRefreshable.update(UPDATED_VALUE);
+        container.close();
+
+        verify(mockCloseableExtra).close();
+    }
+
+    @Test
+    public void noFurtherResourceCreatedAfterClose() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        container.close();
+        settableRefreshable.update(UPDATED_VALUE);
+
+        verify(closeableFactory, times(1)).apply(any());
+    }
+
+    @Test
+    public void runWithResourceFailsAfterClose() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        container.close();
+
+        assertThatThrownBy(() -> container.runWithResource(_closeable -> null))
+                .isInstanceOf(SafeIllegalStateException.class);
+    }
+
+    @Test
+    public void closeReleasesLocksAfterException() throws IOException {
+        doThrow(new SafeIllegalStateException()).when(mockCloseable).close();
+
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        assertThatThrownBy(container::close).isInstanceOf(SafeIllegalStateException.class);
+
+        Future<?> future = executor.submit(container::close);
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(future::isDone);
```

Comment:
ok you're going to have to explain this bit of magic to me - why do we need a future here?